src/threaddeadlock.cpp

```diff
@@ -0,0 +1,718 @@
+        if (isExclusive)
+        {
+            auto it = lockdata.locksheldbythread.find(tid);
+            if (it == lockdata.locksheldbythread.end() || it->second.empty())
+            {
+                // intentionally left blank
+            }
+            else
+            {
+                for (auto &lockStackLock : it->second)
+                {
+                    // if we have a lock and it isnt exclusive and we are attempting to get exclusive
+                    // then we will deadlock ourself
+                    if (lockStackLock.first == c && lockStackLock.second.GetExclusive() == false)
+                    {
+                        self_deadlock_detected(now, lockStackLock);
+                    }
+                }
+            }
+        }
+        else
+        {
+            // intentionally left blank
+        }
+    }
+    else if (type == LockType::RECURSIVE)
+    {
+        // this lock can not deadlock itself
+        // intentionally left blank
+    }
+
+    // Begin general deadlock checks for all lock types
+    AddNewLock(now, tid);
+    AddNewWaitingLock(c, tid, isExclusive);
+
+    // if we have exclusive lock(s) and we arent requesting an exclusive lock...
+    if (!isExclusive)
+    {
+    TEST_5:
+    TEST_8:
+        std::vector<LockStackEntry> deadlocks;
+        std::set<uint64_t> threads;
+        // then we can only deadlock if we are locking a thread that is currently held in exclusive state by someone
+        // else
+        if (ReadRecursiveCheck(tid, c, tid, c, true, deadlocks, threads))
+        {
+            // we have a deadlock where we are requesting shared ownership on a mutex that is exclusively owned by
+            // another thread which has either a shared or exlcusive request on a mutex we have exclusive ownership over
+            potential_deadlock_detected(now, deadlocks, threads);
+        }
+    }
+    // if we have exclusive lock(s) and we are requesting another exclusive lock
+    if (isExclusive)
+    {
+    TEST_6:
+    TEST_7:
+        std::vector<LockStackEntry> deadlocks;
+        std::set<uint64_t> threads;
+        if (WriteRecursiveCheck(tid, c, tid, c, true, deadlocks, threads))
+        {
+            potential_deadlock_detected(now, deadlocks, threads);
+        }
+    }
+}
+
+// removes all instances of the critical section
+void DeleteCritical(void *cs)
+{
+    // remove all instances of the critical section from lockdata
+    std::lock_guard<std::mutex> lock(lockdata.dd_mutex);
+    if (!lockdata.available)
+    {
+        // lockdata was already deleted
+        return;
+    }
+    if (lockdata.readlockswaiting.count(cs))
+        lockdata.readlockswaiting.erase(cs);
+    if (lockdata.writelockswaiting.count(cs))
+        lockdata.writelockswaiting.erase(cs);
+    if (lockdata.readlocksheld.count(cs))
+        lockdata.readlocksheld.erase(cs);
+    if (lockdata.writelocksheld.count(cs))
+        lockdata.writelocksheld.erase(cs);
```

Comment:
You don't need to check if it exists in the map first, you can just call erase even if the key doesn't exist

Reply:
you are right, the only exception in erase (value) is one that can be thrown by the compare object. will fix